### DISCUSSION TITLE
Subset equivalency assertion

### DISCUSSION THREAD

**Discussion by rcdailey** [fluentassertions/fluentassertions#1547] [Q&A]
### Description

I have a container (superset) with 3 strings in it. I need to make sure that 2 strings are present in that container. Usually this comparison requires doing either a substring search or a case insensitive comparison. In either case, this involves specifying a lambda predicate for the test in the assertion.

`BeSubsetOf()` exists, but this requires the equals method be implemented on the objects. It would be ideal to have a similar method for testing for equivalency, especially when the subset check involves complex objects. Something like `BeEquivalentSubsetOf()`?

### Complete minimal example reproducing the issue

Below is an example of how I execute my test today.

```cs
var ctx = TestCorrelator.GetLogEventsFromCurrentContext().ToList();
foreach (var log in expectedLogs)
{
    ctx.Should().Contain(evt => evt.MessageTemplate.Text.Contains(log));
}
```

### Expected behavior:

What would be ideal is to do something like this:

```cs
var myStrings = new List<string> { "Hello World", "Foo Bar", "No Match" };
var expectedSubstrings = new List<string> { "Foo", "World" };
expectedSubstrings.Should().BeEquivalentSubsetOf(myStrings, (a, b) => b.Contains(a));
```

Using my earlier example:

```cs
expectedLogs.Should().BeEquivalentSubsetOf(TestCorrelator.GetLogEventsFromCurrentContext(),
    (log, evt) => evt.MessageTemplate.Text.Contains(log));
```

### Versions

Using version `5.*` with .NET 5

**Comment by dennisdoomen** [comment]
From what I can see, you're looking for a very specific use case that will most likely be very niche. The closest I can think of is the `Should().SatisfyRespectively` or using the `Using`/`WhenTypeIs` options of `BeEquivalentTo` like in [this example](https://github.com/fluentassertions/fluentassertions/blob/6517b442d048010e0acbf1d6174f08b167be917c/Tests/FluentAssertions.Specs/Equivalency/CollectionEquivalencySpecs.cs#L1380).

**Comment by rcdailey** [comment]
I find it interesting you feel that this is niche. You obviously have observed a wider market / set of use cases than I have. I personally feel like I run into this a lot. Maybe it has to do with me not really approaching my unit tests correctly.

For example, I want to verify that a particular method I call logs a warning. There's other logs in there I don't care about, like debug and informational logs. But I need to make sure one or two warnings appear in that list.

Another example, I have a method that parses a markdown document. For each header in the document, it parses out the header and certain keywords out of any paragraphs under it. In my test, I only want to check that two or more specific headers got parsed and ignore the rest. Furthermore, the dataset returned is a list of some complex ref type. The test should only pass if there exists 2 objects in the list that match a specific subset, and that subset may only look at specific properties on each object (not the full object graph).

If you were to tell me these are pretty sophisticated tests, I'd agree with you. But niche is something I can't really comment on. After all I only see what is right in front of me.

I'm hoping you can tell me I'm doing something wrong here. Because if not, it will continue to be a bit of a struggle to find a readable and convenient means to perform these types of customized subset assertions.

Thanks for your quick response.

**Comment by dennisdoomen** [comment]
> For example, I want to verify that a particular method I call logs a warning. There's other logs in there I don't care about, like debug and informational logs. But I need to make sure one or two warnings appear in that list.

For those, you can use the overload of `collection.Should().Contain()` that takes a lambda expression, right?

> Furthermore, the dataset returned is a list of some complex ref type. The test should only pass if there exists 2 objects in the list that match a specific subset, and that subset may only look at specific properties on each object (not the full object graph).

I guess you could use two calls of `collection.Should().ContainEquivalentOf` for that one. 

> I'm hoping you can tell me I'm doing something wrong here. 

You're not doing anything wrong here. There are so many different ways of testing. I'm already happy that you are writing unit tests in the first place. And to be fair, I regularly run into scenarios where I wonder if there could not be a simpler way to assert some complicated scenario. But often, either I _can_ come up with a syntax that would be incredibly hard to explain to others, or I simply can't find anything that the compiler would be able to handle.

**Comment by jnyrup** [comment]
Here's an example of using `ContainEquivalentOf` and case-insensitive substring matching.
The downside of this example is that it does not take into account that an item in `subject` could match multiple items in `expected`.

```cs
private class LogEvent
{
    public LogEvent(string text) => MessageTemplate = new() { Text = text };

    public MessageTemplate MessageTemplate { get; set; }

    public int IgnoreMe { get; set; }
}

private class MessageTemplate
{
    public string Text { get; set; }

    public int AlsoIgnoreMe { get; set; }
}

[Fact]
public void ComplexObjects()
{
    var subject = new LogEvent[]
    {
        new("...Foo..."),
        new("...bAr..."),
        new("...baZ...")
    };

    var expected = new[]
    {
        new { MessageTemplate = new { Text = "foo" } },
        new { MessageTemplate = new { Text = "bar" } },
    };

    foreach (var item in expected)
    {
        subject.Should().ContainEquivalentOf(item, opt => opt
            .Using<string>(ctx => ctx.Subject.Should().ContainEquivalentOf(ctx.Expectation))
            .WhenTypeIs<string>());
    }
}
```

If  you want a more readable test body (which is often a desirable goal for unit tests), you could make an extension method

```cs
[Fact]
public void ComplexObjectsCustom()
{
    var subject = new LogEvent[]
    {
        new("...Foo..."),
        new("...bAr..."),
        new("...baZ...")
    };

    var expected = new[]
    {
        new { MessageTemplate = new { Text = "foo" } },
        new { MessageTemplate = new { Text = "bar2" } },
    };

    subject.Should().ContainEquivalentsOf(expected, opt => opt
        .Using<string>(ctx => ctx.Subject.Should().ContainEquivalentOf(ctx.Expectation))
        .WhenTypeIs<string>());
}
```

```cs
internal static class GenericCollectionAssertionExtensions
{
    [CustomAssertion]
    public static AndConstraint<TAssertions> ContainEquivalentsOf<TSubject, TAssertions, TExpectation>(
        this CollectionAssertions<TSubject, TAssertions> parent,
        IEnumerable<TExpectation> expectations,
        Func<EquivalencyAssertionOptions<TExpectation>, EquivalencyAssertionOptions<TExpectation>> config,
        string because = "",
        params object[] becauseArgs)
            where TSubject : IEnumerable
            where TAssertions : CollectionAssertions<TSubject, TAssertions>
    {
        using var scope = new AssertionScope();

        foreach (TExpectation expectation in expectations)
        {
            parent.ContainEquivalentOf(expectation, config);
        }

        string[] failures = scope.Discard();

        Execute.Assertion
            .ForCondition(!failures.Any())
            .BecauseOf(because, becauseArgs)
            .FailWith("Expected {context:collection} {0} to contain equivalents of {1}{reason}.", parent.Subject, expectations);

        return new((TAssertions)parent);
    }
}
```

When FluentAssertions 6.0 is released, it can be improved a bit to also return the matched equivalents.
```cs
internal static class GenericCollectionAssertionExtensions
{
    [CustomAssertion]
    public static AndWhichConstraint<TAssertions, IEnumerable<T>> ContainEquivalentsOf<TCollection, T, TAssertions, TExpectation>(
        this GenericCollectionAssertions<TCollection, T, TAssertions> parent,
        IEnumerable<TExpectation> expectations,
        Func<EquivalencyAssertionOptions<TExpectation>, EquivalencyAssertionOptions<TExpectation>> config,
        string because = "",
        params object[] becauseArgs)
            where TCollection : IEnumerable<T>
            where TAssertions : GenericCollectionAssertions<TCollection, T, TAssertions>
    {
        using var scope = new AssertionScope();

        var equivalents = new List<T>();

        foreach (TExpectation expectation in expectations)
        {
            T equivalent = parent.ContainEquivalentOf(expectation, config).Which;
            equivalents.Add(equivalent);
        }

        string[] failures = scope.Discard();

        Execute.Assertion
            .ForCondition(!failures.Any())
            .BecauseOf(because, becauseArgs)
            .FailWith("Expected {context:collection} {0} to contain equivalents of {1}{reason}.", parent.Subject, expectations);

        return new((TAssertions)parent, equivalents);
    }
}
```

Once `BeEquivalentTo` and friends are adapted to use our new maximum matching algorithm (see #1486), it should be fairly simple to implement a `ContainEquivalentsOf(IEnumerable<TExpectation> subset)`.

**Comment by rcdailey** [comment]
> > For example, I want to verify that a particular method I call logs a warning. There's other logs in there I don't care about, like debug and informational logs. But I need to make sure one or two warnings appear in that list.
> 
> For those, you can use the overload of `collection.Should().Contain()` that takes a lambda expression, right?

Sorry I've unintentionally over-simplified my example. But what I need is to be able to specify the expected container *and* a lambda. Right now that option doesn't seem to exist:

![image](https://user-images.githubusercontent.com/1768054/116788969-1f1e3880-aa72-11eb-970e-7e9b3b765eee.png)

The current version that takes a lambda will stop after the first match. I need it to continue evaluating the lambda until the number of `true` returns count matches the size of the expected container.

I remember looking at `HasCount()` to do this, which does take a lambda, but it passes in a count not each element. So it wouldn't work either.

> > Furthermore, the dataset returned is a list of some complex ref type. The test should only pass if there exists 2 objects in the list that match a specific subset, and that subset may only look at specific properties on each object (not the full object graph).
> 
> I guess you could use two calls of `collection.Should().ContainEquivalentOf` for that one.

Yeah that's essentially what I'm doing now, except with a `foreach` instead of unrolling the loop. It matters more with larger data sets. My examples here are simplified for discussion purposes but I think you get the idea. It's all about that extra sugar coating. I've started to view explicit `foreach` in a test as "bad" since it isn't fluent. Maybe that's self-imposed and wrong of me to think that way.

> > I'm hoping you can tell me I'm doing something wrong here.
> 
> You're not doing anything wrong here. There are so many different ways of testing. I'm already happy that you are writing unit tests in the first place. And to be fair, I regularly run into scenarios where I wonder if there could not be a simpler way to assert some complicated scenario. But often, either I _can_ come up with a syntax that would be incredibly hard to explain to others, or I simply can't find anything that the compiler would be able to handle.

Thanks, I needed this reassurance. It's definitely worth mentioning (and I should have made this clear at the start if I forgot to) that there's no real issue here. Fluent Assertions is an amazing library that already makes things flow in a more human readable way. At the scale of problems it is solving, this is really just a nitpick.

Part of this was me feeling like I wasn't fitting into the fluent model very well. But you're right there's a lot of variations. A single foreach, at the end of the day, isn't a big deal.

I very much value and appreciate your feedback. Thank you very much.

**Comment by dennisdoomen** [comment]
Thank you for taking the time to suggest improvements.

**Comment by rcdailey** [comment]
@jnyrup I love the examples, thank you. Your idea certainly offers some inspiration especially for the more complicated scenarios I talked about. I do wonder then if `BeEquivalentsOf()` ultimately replaces `BeSubsetOf()`? In a way, the bad thing about me asking for too much is that it blurs the line between the purpose of certain methods. The reason I was drawn to (and focused on) the `BeSubsetOf()` method is because when I think of solving this problem and search for solutions to it, I use the term "subset". It could be that instead of plural form `Equivalents` (which initially I thought was a typo), using `EquivalentSubset` might make more sense (and ultimately be the thing I'm asking for here). Seems much easier to debate what to name a function than to ask for niche functionality :)

I'm happy to close this issue as I feel like reasonably close solutions to what I wanted have been explained here. Only reason I'm not closing it right away is because it might serve as a ticket to track the idea @jnyrup shared. Either way, if you'd like to close this I am perfectly fine with that. Thanks again to you both.

**Comment by meeple142** [comment]
:) I spent an hour or two trying to find an `EquivalentSubset`  function last week too! I really love the power of `BeEquivalentTo` to match complex ref types in two arrays through the object graph comparison (and the great fail messages). But I had a test where the two lists were not the same size. I saw the `BeSubsetOf` and thought that it would use the object graph comparison too.  
Like rcdailey  I am new to FA (and I would add "unit testing in general"), so I thought that it must be me thinking about this wrong. I ended up with a `LINQ.GroupJoin` and a `for` loop with non-container FA assertions. I will definitely look at the other  suggestions. Thank you!

My use case is an integration test not a unit test so maybe that's why I can see needing `BeEquivalentSubsetOf` often?

I am writing an REST API wrapper to communicate with a 3rd party. We are using Refit, Polly and Newtonsoft.json. Unfortunately the 3rd party's documentation needs some work and the endpoints return many undocumented properties. Also they change things often without notification. I am writing tests to make sure that the return type classes we have created match the Json they send back. I plan to run the tests often to make sure we get notified if/when something breaks. So to test, I make a call with our library and then a "raw call" that returns a Newtonsoft.Json JToken. I then map the props of "our" object to a `List<CustomClass>` and then map the props of the JToken to a `List<CustomClass>`. Then I wanted to do something like
```c#
ours.Should().BeEquivalentSubsetOf(JTokenPropList)
```
and have it act like `BeEquivalentOf` and only pass if it finds an equivalent match in expected for every item in subject array and not worry if the subject array length is less than the expected array length.  But as pointed out above, this doesn't handle potential duplicates in expected. Potentially not a big concern in unit tests because you control expected but in an integration test it shouldn't be over looked. So maybe this is more "niche" than I thought.

Any suggestions different than what's listed above?
Thank you very much!!! This is a wonderful library!

**Comment by dennisdoomen** [comment]
@meeple142 maybe [this is what you need](https://github.com/dennisdoomen/EffectiveTddDemo/blob/master/Tests/DocumentManagement.Specs/13_SimplerDeserialization/StatisticsControllerSpecs.cs#L80)? It's using a [custom extension method](https://github.com/dennisdoomen/EffectiveTddDemo/blob/master/Tests/DocumentManagement.Specs/13_SimplerDeserialization/HttpClientExtensions.cs#L25) that tries to deserialize the JSON into an anonymous type defined by an existing anonymous type and then uses FA to compare that anonymous type with itself.
